nemo_text_processing/inverse_text_normalization/fr/graph_utils.py

```diff
@@ -0,0 +1,235 @@
+    from pynini.examples import plurals
+    from pynini.lib import byte, pynutil, utf8
+
+    NEMO_CHAR = utf8.VALID_UTF8_CHAR
+
+    NEMO_DIGIT = byte.DIGIT
+    NEMO_LOWER = pynini.union(*string.ascii_lowercase).optimize()
+    NEMO_UPPER = pynini.union(*string.ascii_uppercase).optimize()
+    NEMO_ALPHA = pynini.union(NEMO_LOWER, NEMO_UPPER).optimize()
+    NEMO_ALNUM = pynini.union(NEMO_DIGIT, NEMO_ALPHA).optimize()
+    NEMO_HEX = pynini.union(*string.hexdigits).optimize()
+    NEMO_NON_BREAKING_SPACE = u"\u00A0"
+    NEMO_SPACE = " "
+    NEMO_WHITE_SPACE = pynini.union(" ", "\t", "\n", "\r", u"\u00A0").optimize()
+    NEMO_NOT_SPACE = pynini.difference(NEMO_CHAR, NEMO_WHITE_SPACE).optimize()
+    NEMO_NOT_QUOTE = pynini.difference(NEMO_CHAR, r'"').optimize()
+
+    NEMO_PUNCT = pynini.union(*map(pynini.escape, string.punctuation)).optimize()
+    NEMO_GRAPH = pynini.union(NEMO_ALNUM, NEMO_PUNCT).optimize()
+
+    NEMO_SIGMA = pynini.closure(NEMO_CHAR)
+
+    delete_space = pynutil.delete(pynini.closure(NEMO_WHITE_SPACE))
+    insert_space = pynutil.insert(" ")
+    delete_extra_space = pynini.cross(pynini.closure(NEMO_WHITE_SPACE, 1), " ")
+
+    # French frequently compounds numbers with hyphen.
+    delete_hyphen = pynutil.delete(pynini.closure("-", 0, 1))
+    insert_hyphen = pynutil.insert("-")
+
+    suppletive = pynini.string_file(get_abs_path("data/suppletive.tsv"))
+
+    _s = NEMO_SIGMA + pynutil.insert("s")
+    _x = NEMO_SIGMA + pynini.string_map([("eau"), ("eu"), ("ou")]) + pynutil.insert("x")
+    _aux = NEMO_SIGMA + pynini.string_map([("al", "aux"), ("ail", "aux")])
+
+    graph_plural = plurals._priority_union(
+        suppletive, plurals._priority_union(_s, pynini.union(_x, _aux), NEMO_SIGMA), NEMO_SIGMA
+    ).optimize()
+
+    SINGULAR_TO_PLURAL = graph_plural
+    PLURAL_TO_SINGULAR = pynini.invert(graph_plural)
+    TO_LOWER = pynini.union(*[pynini.cross(x, y) for x, y in zip(string.ascii_uppercase, string.ascii_lowercase)])
+    TO_UPPER = pynini.invert(TO_LOWER)
+
+    PYNINI_AVAILABLE = True
+except (ModuleNotFoundError, ImportError):
+    # Create placeholders
+    NEMO_CHAR = None
+
+    NEMO_DIGIT = None
+    NEMO_LOWER = None
+    NEMO_UPPER = None
+    NEMO_ALPHA = None
+    NEMO_ALNUM = None
+    NEMO_HEX = None
+    NEMO_NON_BREAKING_SPACE = u"\u00A0"
+    NEMO_SPACE = " "
+    NEMO_WHITE_SPACE = None
+    NEMO_NOT_SPACE = None
+    NEMO_NOT_QUOTE = None
+
+    NEMO_PUNCT = None
+    NEMO_GRAPH = None
+
+    NEMO_SIGMA = None
+
+    delete_space = None
+    insert_space = None
+    delete_extra_space = None
+
+    delete_hyphen = None
+    insert_hyphen = None
+
+    suppletive = None
+    _s = None
+    _x = None
+    _aux = None
+
+    graph_plural = None
+
+    SINGULAR_TO_PLURAL = None
+    PLURAL_TO_SINGULAR = None
+    TO_LOWER = None
+    TO_UPPER = None
+
+    PYNINI_AVAILABLE = False
+
+
+def generator_main(file_name: str, graphs: Dict[str, pynini.FstLike]):
+    """
```

Comment:
Is this function and the functions below differ from the same functions in the en/ folder? Can we avoid duplication?

Reply:
They're all the same functions except for the delete/insert_hyphen and plurals transformer. We can just import the rest from the english one, but it seems easier just to have a duplicate here in case further uses come up down the road. (Also would keep import statements uniform.)